### PR TITLE
command: add tkctl diagnose subcommand to export diagnostic information

### PR DESCRIPTION
**Pull request by qiffang**
### What problem does this PR solve? 
Currently, It is not convenient to export diagnostic data. Though we can view the information by kubernetes command, viewing these informations is tedious. So we need to provide a similar tool to export diagnostic data.

This PR adds a diagnose subcommand for `tkctl` which can export these information:
- TidbCluster
- StatefulSet
- PVC
- Service
- ConfigMap
- Pod
- Log

### What is changed and how does it work?
The `diagnose` command export data from kubernetes and render data like `kubectl get -o wide` command. For container log,  it use a stream to the log file which can be piped directly to the response. This avoids out of memory issues. It also read archived logs created by log rotation or container crash.

### Check List 

Tests 

- Manual test

```
./tkctl diagnose
Error: required flag(s) "path" not set


Examples:

		# specify a tidb cluster to use
		tkctl diagnose

		# diagnose specify tidb cluster information
		tkctl diagnose -t demo-cluster

Options:
      --byteReadLimit=500000: The maximum number of bytes dump log.
      --path='': The log path to dump.
      --since=1h0m0s: Return logs newer than a relative duration like 1m, or 3h.

Usage:
  tkctl diagnose [flags] [options]

Use "tkctl options" for a list of global command-line options (applies to all commands).

required flag(s) "path" not set
```

```
./tkctl diagnose --path=/tmp/test --since 5h

The result is here
$dumppath
├── pod-logs
│   ├── $name-pd-0-pd-$namespace.log
│   ├── $name-pd-1-pd-$namespace.log
│   ├── $name-pd-2-pd-$namespace.log
│   ├── $name-tidb-0-tidb-$namespace.log
│   ├── $name-tikv-0-tikv-$namespace.log
│   ├── $name-tikv-1-tikv-$namespace.log
│   └── $name-tikv-2-tikv-$namespace.log
├── pod-infos
│   ├── $name-pd-0-$namespace.yaml
│   ├── $name-pd-1-$namespace.yaml
│   ├── $name-pd-2-$namespace.yaml
│   ├── $name-tidb-0-$namespace.yaml
│   ├── $name-tidb-1-$namespace.yaml
│   ├── $name-tikv-0-$namespace.yaml
│   ├── $name-tikv-1-$namespace.yaml
│   └── $name-tikv-2-$namespace.yaml
├── resources
├── $name-$namespace-statefulsets-info.yaml
├── $name-$namespace-pvc-info.yaml
├── $name-$namespace-service-info.yaml
├── $name-$namespace-configmap-info.yaml
└── $name-$namespace-tidbcluster-info.yaml

```

Code changes
 - Has Go code change

Side effects

 - No

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:
 
 ```release-note
add `tkctl diagnose` subcommand to export diagnostic information.
 ```

### CODE REVIEW

File: pkg/tkctl/cmd/dump/dump.go [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+func NewTiDBClusterStatefulDescriber(internalCli internalclientset.Clientset, tc *v1alpha1.TidbCluster) *tidbClusterStatefulDescriber {
+	return &tidbClusterStatefulDescriber{
+		internalCli: internalCli,
+		tc:          tc,
+	}
+}
+
+// Dump dump the details of statefulsets from a particular tidb cluster.
+func (d *tidbClusterStatefulDescriber) Dump(logPath string) error {
+	logFile, err := os.Create(filepath.Join(logPath, fmt.Sprintf("%s-%s-statefulsets-info.log", d.tc.Name, d.tc.Namespace)))
```
Comment (weekface) [inline pkg/tkctl/cmd/dump/dump.go]:
```suggestion
	logFile, err := os.Create(filepath.Join(logPath, fmt.Sprintf("%s-%s-statefulsets-info.yaml", d.tc.Name, d.tc.Namespace)))
```

File: pkg/tkctl/cmd/dump/dump.go [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+type tidbClusterDescriber struct {
+	internalCli *internalclientset.Clientset
+	dynamicCli  dynamic.Interface
+
+	mapping *meta.RESTMapping
+	tc      *v1alpha1.TidbCluster
+}
+
+// NewTiDBClusterDummper returns a tidbClusterDumper.
+func NewTiDBClusterDescriber(internalCli *internalclientset.Clientset, dynamicCli dynamic.Interface, tc *v1alpha1.TidbCluster) *tidbClusterDescriber {
```
Comment (weekface) [inline pkg/tkctl/cmd/dump/dump.go]: Just print the `TidbCluster` to `yaml` or `json` to a file is ok. No need `describe` them.

File: pkg/tkctl/cmd/dump/dump.go (hunk 1 above)
Comment (qiffang) [reply pkg/tkctl/cmd/dump/dump.go]: updated it

File: pkg/tkctl/cmd/dump/dump.go (hunk 2 above)
Comment (qiffang) [reply pkg/tkctl/cmd/dump/dump.go]: Done

File: pkg/tkctl/cmd/dump/dump.go [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+		controller.TiDBMemberName(d.tc.Name),
+		controller.TiKVMemberName(d.tc.Name),
+		controller.PDMemberName(d.tc.Name),
+	} {
+		ps, err := d.kubeCli.AppsV1().StatefulSets(d.tc.Namespace).Get(sn, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if err = writeString(logFile, "#"+sn+"\n"); err != nil {
```
Comment (weekface) [inline pkg/tkctl/cmd/dump/dump.go]: the exported YAML file doesn't have `apiVersion` and `kind` attribute.

Comment (qiffang) [reply pkg/tkctl/cmd/dump/dump.go]: Fixed.

File: pkg/tkctl/cmd/dump/dump.go [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+	pod.SetGroupVersionKind(v1.SchemeGroupVersion.WithKind("Pod"))
+	return &podDumper{
+		kubeCli:       kubeCli,
+		pod:           pod,
+		sinceSeconds:  sinceSeconds,
+		byteReadLimit: byteReadLimit,
+	}
+}
+
+// Dump dump detail information and logs of pod from a particular pod.
```
Comment (tennix) [inline pkg/tkctl/cmd/dump/dump.go]:
```suggestion
// Dump dumps detail information and logs of pod from a particular pod.
```

File: pkg/tkctl/cmd/dump/dump.go [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+
+	path = fmt.Sprintf("%s%cpod-logs", logPath, filepath.Separator)
+	if err := createPathIfNotExist(path); err != nil {
+		return err
+	}
+
+	return d.DumpLog(path)
+}
+
+// Dump dump the details of a named Pod from a particular namespace.
```
Comment (tennix) [inline pkg/tkctl/cmd/dump/dump.go]:
```suggestion
// Dump dumps the details of a named Pod from a particular namespace.
```

File: pkg/tkctl/cmd/dump/dump.go [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+
+	body, err := yaml.Marshal(d.pod)
+	if err != nil {
+		return err
+	}
+
+	return writeString(logFile, string(body))
+}
+
+// Dump dump dump the logs for the last terminated container and current running container. If info about the container is not available then a specific
```
Comment (tennix) [inline pkg/tkctl/cmd/dump/dump.go]:
```suggestion
// Dump dumps the logs for the last terminated container and current running container. If info about the container is not available then a specific
```

File: pkg/tkctl/cmd/dump/dump.go [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+			return err
+		}
+
+		d.DumpContainerLog(filepath.Join(logPath, fmt.Sprintf("%s-%s-%s-p.log", d.pod.Name, c.Name, d.pod.Namespace)), c.Name, true)
+	}
+
+	return nil
+}
+
+// DumpContainerLog dump logs for particular pod and container. Previous indicates to read archived logs created by log rotation or container crash
```
Comment (tennix) [inline pkg/tkctl/cmd/dump/dump.go]:
```suggestion
// DumpContainerLog dumps logs for particular pod and container. Previous indicates to read archived logs created by log rotation or container crash
```

**Comment by qiffang** [comment]
The `diagnose` to replace `dump` is ok?

**Comment by tennix** [comment]
> @tennix , using `diagnose` to replace `dump` is ok?

Yes, I think so. The `dump` is confused to users since we are running databases.

File: pkg/tkctl/cmd/dump/dump.go [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+	"strings"
+	"time"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap.com/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/client/clientset/versioned"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	"github.com/pingcap/tidb-operator/pkg/label"
+	"github.com/pingcap/tidb-operator/pkg/tkctl/config"
+	"github.com/pingcap/tidb-operator/pkg/tkctl/readable"
+
```
[the remark this answers is not in the document]
Comment (qiffang) [reply pkg/tkctl/cmd/dump/dump.go]: Done

File: pkg/tkctl/cmd/dump/dump.go [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+		Long:    dumpLongDesc,
+		Example: dumpExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(o.Complete(tkcContext, cmd, args))
+			cmdutil.CheckErr(o.Run())
+		},
+	}
+
+	cmd.Flags().StringVar(&o.logPath, "path", "", "The log path to dump.")
+	cmd.Flags().DurationVar(&o.since, "since", 3600000000000, "Return logs newer than a relative duration like 1m, or 3h.")
```
[the remark this answers is not in the document]
Comment (qiffang) [reply pkg/tkctl/cmd/dump/dump.go]: Done

File: pkg/tkctl/cmd/dump/dump.go (hunk 7 above)
Comment (qiffang) [reply pkg/tkctl/cmd/dump/dump.go]: Done

File: pkg/tkctl/cmd/dump/dump.go (hunk 6 above)
Comment (qiffang) [reply pkg/tkctl/cmd/dump/dump.go]: Done

File: pkg/tkctl/cmd/dump/dump.go (hunk 5 above)
Comment (qiffang) [reply pkg/tkctl/cmd/dump/dump.go]: Done

File: pkg/tkctl/cmd/dump/dump.go (hunk 4 above)
Comment (qiffang) [reply pkg/tkctl/cmd/dump/dump.go]: Done

**Comment by tennix** [comment]
@qiffang Please fix the conflict.

**Pull request merged**